allensdk/brain_observatory/behavior/write_nwb/nwb_writer_utils.py

```diff
@@ -0,0 +1,157 @@
+                    f"comparing session object metadata and data from "
+                    "the behavior stimulus pickle file.\n"
+                    f"\tObject data={bs_val};\n"
+                    f"\tStim data={stim_value}\n"
+                )
+    if len(error_message) > 0:
+        raise ValueError(error_message)
+
+
+class NBWriterCompareMetadata(NWBWriter):
+    """Writer with additional checks against the input metadata.
+    """
+
+    def write_nwb(self,
+                  id_column_name: str = 'behavior_session_id',
+                  skip_metadata: list = [],
+                  skip_stim: list = [],
+                  **kwargs):
+        """Tries to write nwb to disk. If it fails, the filepath has ".error"
+        appended
+
+        Parameters
+        ----------
+        kwargs: kwargs sent to `from_nwb`, `to_nwb`
+
+        """
+        from_lims_kwargs = {
+            k: v for k, v in kwargs.items()
+            if k in inspect.signature(self._serializer.from_lims).parameters}
+        lims_session = self._serializer.from_lims(
+            self._session_data[id_column_name],
+            **from_lims_kwargs)
+        lims_session = self._modify_session(lims_session)
+
+        try:
+            nwbfile = self._write_nwb(
+                session=lims_session, **kwargs)
+            _compare_metadata(
+                input_id=self._session_data[id_column_name],
+                input_session=lims_session,
+                table_metadata=self._session_data,
+                skip_metadata=skip_metadata,
+                skip_stim=skip_stim,
+            )
+            self._compare_sessions(nwbfile=nwbfile, lims_session=lims_session,
+                                   **kwargs)
+            os.rename(self.nwb_filepath_inprogress, self._nwb_filepath)
+        except Exception as e:
+            if os.path.isfile(self.nwb_filepath_inprogress):
+                os.rename(self.nwb_filepath_inprogress,
+                          self._nwb_filepath_error)
+            raise e
+
+    def _modify_session(self, lims_session: BehaviorSession):
+        """Edit data values in the session with data from the session_data
+        dict.
+
+        In some cases, the values of certain metadata is defined by selection
+        of session released. Since the session objects do not know this
+        selection of sessions we need to modify their internal values for
+        consistency. For example, we only want to calculate the
+        targeted_imaging_depth from released experiments so copy the value
+        from the table which knows which experiments are being released and
+        which aren't.
+
+        Parameters
+        ----------
+        lims_session : BehaviorSession
+            Input session to modify.
+
+        Returns
+        -------
+        output_session : BehaviorSession
+            Session with modifyed values.
+        """
+        if "targeted_imaging_depth" in self._session_data.keys():
```

Comment:
This should be in an ophys-specific class. Recommend to move to the ophys nwb writer.